experimental/ast/codec.go

```diff
@@ -0,0 +1,469 @@
+			kind = compilerpb.Def_KIND_EXTEND
+		case DefKindField:
+			kind = compilerpb.Def_KIND_FIELD
+		case DefKindEnumValue:
+			kind = compilerpb.Def_KIND_ENUM_VALUE
+		case DefKindOneof:
+			kind = compilerpb.Def_KIND_ONEOF
+		case DefKindGroup:
+			kind = compilerpb.Def_KIND_GROUP
+		case DefKindMethod:
+			kind = compilerpb.Def_KIND_METHOD
+		case DefKindOption:
+			kind = compilerpb.Def_KIND_OPTION
+		}
+
+		proto := &compilerpb.Def{
+			Kind:          kind,
+			Name:          c.path(decl.Name()),
+			Value:         c.expr(decl.Value()),
+			Options:       c.options(decl.Options()),
+			Span:          c.span(decl),
+			KeywordSpan:   c.span(decl.Keyword()),
+			EqualsSpan:    c.span(decl.Equals()),
+			SemicolonSpan: c.span(decl.Semicolon()),
+		}
+
+		if kind == compilerpb.Def_KIND_FIELD || kind == compilerpb.Def_KIND_UNSPECIFIED {
+			proto.Type = c.type_(decl.Type())
+		}
+
+		if signature := decl.Signature(); !signature.Nil() {
+			proto.Signature = &compilerpb.Def_Signature{
+				Span:        c.span(signature),
+				InputSpan:   c.span(signature.Inputs()),
+				ReturnsSpan: c.span(signature.Returns()),
+				OutputSpan:  c.span(signature.Outputs()),
+			}
+
+			signature.Inputs().Iter(func(_ int, t TypeAny) bool {
+				proto.Signature.Inputs = append(proto.Signature.Inputs, c.type_(t))
+				return true
+			})
+			signature.Outputs().Iter(func(_ int, t TypeAny) bool {
+				proto.Signature.Outputs = append(proto.Signature.Outputs, c.type_(t))
+				return true
+			})
+		}
+
+		if body := decl.Body(); !body.Nil() {
+			proto.Body = &compilerpb.Decl_Body{
+				Span: c.span(decl.Body()),
+			}
+			body.Iter(func(_ int, d DeclAny) bool {
+				proto.Body.Decls = append(proto.Body.Decls, c.decl(d))
+				return true
+			})
+		}
+
+		return &compilerpb.Decl{Decl: &compilerpb.Decl_Def{Def: proto}}
+
+	default:
+		panic(fmt.Sprintf("typeToProto: unknown DeclKind: %d", k))
+	}
+}
+
+func (c *codec) options(options CompactOptions) *compilerpb.Options {
+	if options.Nil() {
+		return nil
+	}
+	defer c.checkCycle(options)()
+
+	proto := &compilerpb.Options{
+		Span: c.span(options),
+	}
+
+	options.Iter(func(_ int, o Option) bool {
+		proto.Entries = append(proto.Entries, &compilerpb.Options_Entry{
+			Path:       c.path(o.Path),
+			Value:      c.expr(o.Value),
+			EqualsSpan: c.span(o.Equals),
+		})
+		return true
+	})
+
+	return proto
+}
+
+func (c *codec) expr(expr ExprAny) *compilerpb.Expr {
+	if expr.Nil() {
+		return nil
+	}
+	defer c.checkCycle(expr)()
+
+	switch k := expr.Kind(); k {
+	case ExprKindLiteral:
+		expr := expr.AsLiteral()
+
+		proto := &compilerpb.Expr_Literal{
+			Span: c.span(expr),
+		}
+		if v, ok := expr.Token.AsInt(); ok {
+			proto.Value = &compilerpb.Expr_Literal_IntValue{IntValue: v}
+		} else if v, ok := expr.Token.AsFloat(); ok {
+			proto.Value = &compilerpb.Expr_Literal_FloatValue{FloatValue: v}
+		} else if v, ok := expr.Token.AsString(); ok {
+			proto.Value = &compilerpb.Expr_Literal_StringValue{StringValue: v}
+		}
```

Comment:
Does this not warrant a final `else` that complains about an unknown type of literal (to be symmetric with the default cases in the handful of switches in this file)?